### PR TITLE
[IMP] mail: introduce WebClientViewAttachmentView

### PR DESCRIPTION
**Pull request by seb-odoo**
This will allow using the component directly into the new Owl form.

task-2886642

https://github.com/odoo/enterprise/pull/29182

### CODE REVIEW

File: addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
+/** @odoo-module **/
+
+import { useModels } from '@mail/component_hooks/use_models';
+// ensure components are registered beforehand.
+import '@mail/components/web_client_view_attachment_view/web_client_view_attachment_view';
+import { replace } from '@mail/model/model_field_command';
+import { getMessagingComponent } from "@mail/utils/messaging_component";
```
Comment (oomsveta) [inline addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js:7]: single quotes

File: addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                this.webClientViewAttachmentView.delete();
+            }
+            this.webClientViewAttachmentView = undefined;
+        }
+    }
+
+    /**
+     * @private
+     */
+    async insertFromProps(props) {
```
Comment (oomsveta) [inline addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js:55]: why not prefixing it with _ if it's supposed to be private?

File: addons/mail/static/src/models/thread.js [hunk 3]
```diff
@@ -579,6 +579,7 @@ registerModel({
                 });
             }
             this.update(values);
+            this.messaging.messagingBus.trigger('o-thread-loaded-data', { thread: this });
```
Comment (oomsveta) [inline addons/mail/static/src/models/thread.js:582]: This doesn't seem to be used outside of `test_utils`. How about adding it via an override of thread in tests assets in order not to have it if not needed?

File: addons/mail/static/src/models/web_client_view_attachment_view.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
+/** @odoo-module **/
+
+import { registerModel } from '@mail/model/model_core';
+import { attr, one } from '@mail/model/model_field';
+import { OnChange } from '@mail/model/model_onchange';
+
+import { hidePDFJSButtons } from '@web/legacy/js/libs/pdfjs';
```
Comment (oomsveta) [inline addons/mail/static/src/models/web_client_view_attachment_view.js:7]: single quotes

File: addons/mail/static/src/models/web_client_view_attachment_view.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            }
+            this.component.trigger('preview_attachment_validation');
+        },
+        async setMainAttachmentFromIndex(index) {
+            await this.thread.setMainAttachment(this.thread.attachmentsInWebClientView[index]);
+        },
+        /**
+         * @private
+         */
+        _onChangeThreadAttachmentsInWebClientView() {
```
Comment (oomsveta) [inline addons/mail/static/src/models/web_client_view_attachment_view.js:41]: why are some handlers private while others aren't?

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js [hunk 6]
```diff
@@ -23,6 +24,10 @@ FormRenderer.include({
      */
     init(parent, state, params) {
         this._super(...arguments);
+        this.modelsListener = new Listener({
+            name: `interchangeChatter`,
```
Comment (oomsveta) [inline addons/mail/static/src/widgets/form_renderer/form_renderer.js:28]: why use backticks?

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js [hunk 7]
```diff
@@ -226,40 +246,15 @@ FormRenderer.include({
...
-                this.attachmentViewer.thread !== thread ||
-                this.attachmentViewer.attachments.length !== thread.attachmentsInWebClientView.length
-            ) {
-                this.attachmentViewer.updateContents(thread);
+    async _updateWebClientViewAttachmentViewContainer() {
+        const props = this._makeWebClientViewAttachmentViewContainerProps();
+        try {
+            await this.webClientViewAttachmentViewContainer.update(props);
+        } catch (error) {
+            if (error.message !== "Mounting operation cancelled") {
```
Comment (oomsveta) [inline addons/mail/static/src/widgets/form_renderer/form_renderer.js:255]: looks like this error message doesn't exist anymore?

File: addons/mail/static/tests/helpers/mock_server.js [hunk 8]
```diff
@@ -613,6 +617,31 @@ patch(MockServer.prototype, 'mail', {
             return res;
         });
     },
+    /**
+     * Simulates `register_as_main_attachment` on `ir.attachment`.
+     *
+     * @private
+     * @param {integer} ids
+     * @param {boolean} [force=true]
+     * @returns {Object}
```
Comment (oomsveta) [inline addons/mail/static/tests/helpers/mock_server.js:626]: this doesn't return an object

File: addons/mail/static/src/models/web_client_view_attachment_view.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            const index = this.thread.attachmentsInWebClientView.findIndex(attachment => attachment === this.thread.mainAttachment);
+            this.setMainAttachmentFromIndex(index === this.thread.attachmentsInWebClientView.length - 1 ? 0 : index + 1);
+        },
+        /**
+         * @param {MouseEvent} ev
+         */
+        onClickPrevious(ev) {
+            ev.preventDefault();
+            const index = this.thread.attachmentsInWebClientView.findIndex(attachment => attachment === this.thread.mainAttachment);
+            this.setMainAttachmentFromIndex(index === 0 ? this.thread.attachmentsInWebClientView.length - 1 : index - 1);
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/web_client_view_attachment_view.js:27]: the suggestion is much more complex

File: addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js (hunk 2 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/web_client_view_attachment_view_container/web_client_view_attachment_view_container.js:55]: I'll make it private, but private method in JS is an obsolete context, the difference is not very important.

File: addons/mail/static/src/models/thread.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread.js:582]: Not worth the effort of an override, most of the `messagingBus.trigger` are done only for tests and are not made that way.

File: addons/mail/static/src/models/web_client_view_attachment_view.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/web_client_view_attachment_view.js:7]: single quote is correct?

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/src/widgets/form_renderer/form_renderer.js:28]: copied from others where variable was used inside

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js (hunk 7 above)
Comment (seb-odoo) [reply addons/mail/static/src/widgets/form_renderer/form_renderer.js:255]: Good catch, I guess I'll remove both of them.

**Review by oomsveta — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**